tests/test_interpolation.py

```diff
@@ -753,3 +758,251 @@ def test_none_value_in_quoted_string(restore_resolvers: Any) -> None:
+        return s[0:-1]  # drop last character from string `s`
+
+    OmegaConf.register_new_resolver("cast", cast)
+    OmegaConf.register_new_resolver("drop_last", drop_last)
+    OmegaConf.register_new_resolver("identity", lambda x: x)
+
+    cfg = OmegaConf.structured(cfg)
+
+    val = cfg[key]
+    assert val == expected_value
+
+    node = cfg._get_node(key)
+    assert isinstance(node, Node)
+    assert isinstance(node._dereference_node(), expected_node_type)
+
+
+@pytest.mark.parametrize(
+    ("cfg", "key", "expected_error"),
+    [
+        pytest.param(
+            User(name="Bond", age=SI("${cast:str,seven}")),
+            "age",
+            pytest.raises(
+                InterpolationValidationError,
+                match=re.escape(
+                    dedent(
+                        """\
+                        Value 'seven' could not be converted to Integer
+                            full_key: age
+                        """
+                    )
+                ),
+            ),
+            id="type_mismatch_resolver",
+        ),
+        pytest.param(
+            User(name="Bond", age=SI("${name}")),
+            "age",
+            pytest.raises(
+                InterpolationValidationError,
+                match=re.escape(
+                    dedent(
+                        """\
+                        Value 'Bond' could not be converted to Integer
+                            full_key: age
+                        """
+                    )
+                ),
+            ),
+            id="type_mismatch_node_interpolation",
+        ),
+        pytest.param(
+            StructuredWithMissing(opt_num=None, num=II("opt_num")),
+            "num",
+            pytest.raises(
+                InterpolationValidationError,
+                match=re.escape("Non optional field cannot be assigned None"),
+            ),
+            id="non_optional_node_interpolation",
+        ),
+        pytest.param(
+            SubscriptedList(list=SI("${identity:[a, b]}")),
+            "list",
+            pytest.raises(
+                InterpolationValidationError,
+                match=re.escape("Value 'a' could not be converted to Integer"),
+            ),
+            id="list_type_mismatch",
+        ),
+        pytest.param(
+            MissingDict(dict=SI("${identity:{0: b, 1: d}}")),
+            "dict",
+            pytest.raises(
+                InterpolationValidationError,
+                match=re.escape("Key 0 (int) is incompatible with (str)"),
+            ),
+            id="dict_key_type_mismatch",
+        ),
+    ],
+)
+def test_interpolation_type_validated_error(
+    cfg: Any,
+    key: str,
+    expected_error: Any,
+    restore_resolvers: Any,
+) -> None:
+    def cast(t: Any, v: Any) -> Any:
+        return {"str": str, "int": int}[t](v)  # cast `v` to type `t`
+
+    OmegaConf.register_new_resolver("cast", cast)
+    OmegaConf.register_new_resolver("identity", lambda x: x)
```

Comment:
A good exercise is to use a fixture instead of duplicating those two functions (not insisting, but I think it's good to practice those).

Reply:
No that's a good point, done in c2c1c03